src/realm/object-store/schema.cpp

```diff
@@ -136,9 +138,58 @@ void check_for_embedded_objects_loop(Schema const& schema, std::vector<ObjectSch
         }
     }
 }
-} // namespace
 
-void Schema::validate(bool for_sync) const
+} // end anonymous namespace
+
+std::unordered_set<std::string> Schema::get_embedded_object_orphans() const
+{
+    std::unordered_set<std::string> orphans;
+    std::unordered_set<std::string> visitable;
+    for (auto const& object : *this) {
+        if (!object.is_embedded) {
+            for (auto const& prop : object.persisted_properties) {
+                if (prop.type == PropertyType::Object) {
+                    auto it = this->find(prop.object_type);
+                    REALM_ASSERT(it != this->end());
+                    if (it->is_embedded) {
+                        visitable.insert(it->name);
+                    }
+                }
+            }
+        }
+        else {
+            orphans.insert(object.name);
+        }
+    }
+    for (auto it = orphans.begin(); it != orphans.end();) {
+        if (visitable.find(*it) != visitable.end()) {
+            it = orphans.erase(it);
+        }
+        else {
+            ++it;
+        }
+    }
+    // at this point any remaining orphans can only be reached by another embedded object
+    bool changes_to_process = true;
+    while (orphans.size() > 0 && changes_to_process) {
+        changes_to_process = false;
+        for (auto& name : visitable) {
+            auto it = this->find(name);
+            REALM_ASSERT(it != this->end());
+            for (auto const& prop : it->persisted_properties) {
+                if (prop.type == PropertyType::Object && orphans.find(prop.object_type) != orphans.end()) {
+                    visitable.insert(prop.object_type);
+                    orphans.erase(prop.object_type);
+                    changes_to_process = true;
+                }
+            }
+        }
+    }
+    return orphans;
```

Comment:
I found this logic confusing to follow and took a stab at something that I find simpler:

```suggestion
    std::queue<const ObjectSchema*> to_check;
    for (auto& object : *this) {
        if (!object.is_embedded) {
            to_check.push(&object);
        }
    }

    // Perform a breadth-first search of the schema graph to discover all object
    // types which are reachable from any of the root types
    std::unordered_set<const ObjectSchema*> reachable;
    while (!to_check.empty()) {
        auto object = to_check.front();
        reachable.insert(object);
        for (auto& prop : object->persisted_properties) {
            if (prop.type == PropertyType::Object) {
                auto it = find(prop.object_type);
                REALM_ASSERT(it != this->end());
                if (it->is_embedded && reachable.insert(&*it).second) {
                    to_check.push(&*it);
                }
            }
        }
        to_check.pop();
    }

    // Any object types which weren't found above are orphans
    std::unordered_set<std::string> orphans;
    for (auto& object : *this) {
        if (object.is_embedded && !reachable.count(&object)) {
            orphans.push(object->name);
        }
    }
    return orphans;
```

Feel free to reject this if you don't agree.

Reply:
This is much nicer, both in terms of readability and efficiency. Thanks!